dcp/initialization.py

```diff
@@ -0,0 +1,128 @@
+"""
+Dynamically initialize the dcp mode.
+
+Traverses through the dcp-client CJS modules to generate the dcp Python module
+and build a registry of classes along the way which are automatically wrapped.
+
+Author: Will Pringle <will@distributive.network>
+Date: June 2024
+"""
+
+import sys
+from types import ModuleType as Module
+from typing import Callable
+import pythonmonkey as pm
+from .dry import class_manager, aio
+from . import js
+from . import api
+
+# state
+INIT_MEMO = None
+
+
+def _wrap_js(prop_name, prop_ref):
+    """Determines what the js ref is, if class add to class registry."""
+    if isinstance(prop_ref, pm.JSFunctionProxy):
+        # class
+        if js.utils.isclass(prop_ref):
+            new_bfclass = class_manager.reg.find(prop_ref)
+            if new_bfclass is None:
+                new_bfclass = class_manager.wrap_class(prop_ref, name=prop_name)
+                new_bfclass = class_manager.reg.add(new_bfclass)
+            return new_bfclass
+
+        # function
+        else:
+            def fn_wrapper(*args, **kwargs):
+                ret_val = aio.blockify(prop_ref)(*args, **kwargs)
+                return _wrap_js('dynamically_accessed_property', ret_val)
+            return fn_wrapper
+
+    # js object
+    elif isinstance(prop_ref, pm.JSObjectProxy):
+        return class_manager.wrap_obj(prop_ref)
+
+    elif isinstance(prop_ref, pm.null.__class__):
+        return None
+
+    # py dict
+    else:
+        return prop_ref
+
+    return None
+
+
+class _DynamicModule(Module):
+    """Dynamic get attribute for 'hidden' JS properties on a module."""
+    def __init__(self, name, js_module):
+        super().__init__(name)
+        self._js = js_module
+
+    def __getattr__(self, name):
+        if self._js[name] is not None:
+            return _wrap_js(name, self._js[name])
+        return None
+
+
+class AIOModule(Module):
+    """For getting AIO."""
+    def __init__(self, name, js_module):
+        super().__init__(name)
+        self._js = js_module
+
+    def __getattr__(self, name):
+        js_attr = self._js[name]
+
+        if js_attr is None:
+            return None
+
+        if callable(js_attr):
+            return aio.asyncify(js_attr)
```

Comment:
this should probably be replaced with the class in class_manager since it does all of this for us already... think about cleaning this up later...